File: 10. Week 10 - Interaction and Animation/solutions/r/visualization.R
# Visualize Interactive Dashboard using ggplot2 and plotly

if (!require(ggplot2)) install.packages("ggplot2")
if (!require(readr)) install.packages("readr")
if (!require(plotly)) install.packages("plotly")
library(ggplot2)
library(readr)
library(plotly)

data_path <- "../../data/linked_data.csv"

if (file.exists(data_path)) {
  df <- read_csv(data_path, show_col_types = FALSE)
  
  # 1. Create Base ggplot
  # Scatter Plot of Distance vs Pace
  p <- ggplot(df, aes(x = Distance_km, y = Pace_min_km, 
                      text = paste("Date:", Date, "<br>Dist:", Distance_km, "km<br>Pace:", Pace_min_km))) +
    geom_point(aes(color = Pace_min_km), alpha = 0.7) +
    scale_color_gradient(low = "plum", high = "rebeccapurple") +
    labs(
      title = "Interactive Analysis: Distance vs Pace",
      x = "Distance (km)",
      y = "Pace (min/km)"
    ) +
    theme_light()
  
  # 2. Convert to Plotly for Interaction
  interactive_plot <- ggplotly(p, tooltip = "text")
  
  # 3. Save as HTML
  # htmlwidgets::saveWidget requires pandoc, which might not be in path. 
  # We'll use plotly's export if available or just print info.
  # For this exercise, we generate the object.
  # To save: htmlwidgets::saveWidget(interactive_plot, "interactive_dashboard.html")
  
  # We will try to save it if htmlwidgets is available
  if (!require(htmlwidgets)) install.packages("htmlwidgets")
  library(htmlwidgets)
  
  saveWidget(interactive_plot, "interactive_dashboard.html", selfcontained = TRUE)
  print("Interactive Dashboard saved to interactive_dashboard.html")

} else {
  print(paste("Error: File not found at", data_path))
}

File: 10. Week 10 - Interaction and Animation/solutions/python/visualization.py
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import os

def visualize():
    data_path = '../../data/linked_data.csv'
    
    if not os.path.exists(data_path):
        print(f"Data not found at {data_path}")
        return

    df = pd.read_csv(data_path)
    
    # Setup Figure with Subplots
    fig = make_subplots(
        rows=1, cols=2,
        column_widths=[0.6, 0.4],
        subplot_titles=("Distance vs Pace (Select here)", "Distance by Month")
    )
    
    # 1. Scatter Plot
    scatter = go.Scatter(
        x=df['Distance_km'],
        y=df['Pace_min_km'],
        mode='markers',
        marker=dict(color='#663399', opacity=0.6),
        name='Runs'
    )
    fig.add_trace(scatter, row=1, col=1)
    
    # 2. Bar Chart (Aggregated)
    # Pivot logic isn't dynamic in static Plotly export, but with Dash it would be.
    # For a static HTML export with interaction, Plotly has "selection" events but linked views 
    # usually require Dash or complex JS callbacks.
    # However, standard Plotly simply allows zooming/panning.
    # For this exercise, we'll demonstrate a Dual-View simple dashboard.
    
    # Pre-aggregate for the bar chart
    monthly = df.groupby('Month')['Distance_km'].sum().reset_index()
    
    bar = go.Bar(
        x=monthly['Month'],
        y=monthly['Distance_km'],
        marker_color='#9B59B6',
        name='Monthly Sum'
    )
    fig.add_trace(bar, row=1, col=2)
    
    # Layout
    fig.update_layout(
        title_text="Weekly Distance & Pace Analysis",
        showlegend=False,
        height=500,
        dragmode='select' # Enable box select tool
    )
    
    # Axis labels
    fig.update_xaxes(title_text="Distance (km)", row=1, col=1)
    fig.update_yaxes(title_text="Pace (min/km)", row=1, col=1)
    fig.update_xaxes(title_text="Month", row=1, col=2)
    fig.update_yaxes(title_text="Total Distance", row=1, col=2)
    
    # Save
    fig.write_html("interactive_dashboard.html")
    print("Interactive Dashboard saved to interactive_dashboard.html")
    
    # Ideally, for "Brushing and Linking" in pure Python, one uses:
    # 1. Dash (App framework)
    # 2. Altair (Declarative spec, handles linking natively in HTML)
    # 3. Plotly FigureWidget (Jupyter Notebooks only)
    # Since this is a script producing an artifact, plain HTML output won't perfectly link 
    # (filtering one doesn't update the other without JS).
    # But this meets the requirement of using Plotly for "Interaction" (Zoom/Pan/Hover).

if __name__ == "__main__":
    visualize()
